### ISSUE TITLE
Scalaz 8 legal issues

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz#1729]
I have been informed Scalaz 8.x has been [preliminarily licensed](https://github.com/scalaz/scalaz/issues/1334) under the 7.x license, which appears to be the [3-clause BSD license](https://opensource.org/licenses/BSD-3-Clause).

I would suggest this being a new version, with almost entirely new contributors, it's time to consider a different license. The BSD licenses have no advantages and many disadvantages.

I would unreservedly recommend [LGPL](https://www.gnu.org/licenses/lgpl-java.html), which ensures that users of commercial software based on Scalaz have the right to swap out the dynamically linked library with one of their choosing; and that the software remains free forever.

On a separate issue, some people seem to be under the impression that Scalaz is "public domain" and the source code is not under copyright law. This is explicitly contradicted in the old 7.x license, which [explicitly lists all copyright holders](https://github.com/scalaz/scalaz/blob/series/7.3.x/LICENSE.txt#L1). But it may be worth clarifying this point—that Scalaz 8 is not "public domain", and that copyright has not been assigned to "Scalaz" (since no such legal entity exists), but rather remains with the individual contributors.

I volunteer to do all the work if we reach some sort of consensus on a new license and copyright assignment.

**Comment by jbgi** [comment]
:+1: for LGPLv3

**Comment by aloiscochard** [comment]
I personally don't understand the motivation for the license change.

For a simple mortal than me, could someone list what do we *gains* or *looses* by doing this change?

**Comment by hrhino** [comment]
> could someone list what do we gains or looses by doing this change?

yes please; I'm not very clear on how the various GPL/BSD licenses differ.

**Comment by jdegoes** [comment]
Hopefully @fommil will chime in here, but in brief, and in my view as a contributor to Scalaz 8, there are three compelling reasons to choose LGPL over BSD:

1. LGPL ensures that free software stays free, and that improvements are made available for contribution back into the community. This makes me more likely to donate my time, because I know that the software and all improvements will be free forever.
2. LGPL ensures that users have a choice even in commercial applications. If some company bundles a version of Scalaz into a proprietary application, users have the right to swap that version out with a version of their own choosing, enabling them to fix bugs and tweak as they see fit.
3. LGPL is supported by the weight of the FSF, whereas BSD licenses are essentially legally untested, random copy someone made up for another project a long time ago.

**Comment by aloiscochard** [comment]
@jdegoes okay, this makes sense. Thanks for explaining it clearly.

I realize how important it can be for you, especially point 1.

It seems that people that fear LGPLv3 are just confusing it with the older version, so it's a matter of education I guess?

:+1: for LGPLv3 if it encourage contributions

**Comment by relrod** [comment]
I'm +0 on the license change, but I do tend to personally prefer the more liberal licenses (BSD, MIT, ISC, etc.). While I certainly hope and wish that users of Scalaz contribute any useful changes they make back to the upstream, I don't necessarily feel that's something we should force them to do.

That said, I haven't written any Scalaz 8 code, so I probably don't really have too much of a horse in the race.

> On a separate issue, some people may be under the impression that Scalaz is public domain and the source code is not under copyright law.

I don't understand why you think people would have this impression. There's a `LICENSE.txt` in the root of the repo on the 7 branches and 8 branches that clearly does not waive rights nor copyright interests of the relevant authors.

**Comment by fommil** [comment]
I'm +1 for LGPL. If we want to include code from scalaz7 we need to include their LICENSE (or, infeasibly, ask everybody to relicense, which is not worth it).

That said, I'm not really active in scalaz8 at the moment so I yield all decisions to those pushing the code.

@jdegoes I haven't yet published the ioeffect backport, can you just quickly check that you're happy with the license? I received something saying apache 2.0 but I didn't check it out.

**Comment by hrhino** [comment]
Has anyone used scalaz 7 code in scalaz 8? Everything I've written has been from scratch (although I've used and contributed to scalaz 7 before; not sure if that morally makes a difference)

**Comment by jdegoes** [comment]
@fommil Apache 2 is fine for the backport. 👍 

I think it is feasible to ask everyone who has contributed to _Scalaz 8_ to relicense. Scalaz 7.x, not so much. This might change if any 8 contributors copied over code from 7.

**Comment by aloiscochard** [comment]
I'm pretty confident none of my commits include code copied from Scalaz7 (type-class definition might looks exactly the same, but this is purely incidental... those are so general anyway), also, if code was copied it was from my personal unlicensed/public domain one, so it is totally okay to re-license it.

My guess in that it is the same for the contributions from @jbgi, is that right?

**Comment by jbgi** [comment]
> My guess in that it is the same for the contributions from @jbgi, is that right?

yes!

**Comment by fommil** [comment]
@jdegoes oh it is bsd3, i.e. the scalaz7 license, we can change that for 2.0.0 with the bifunctor to whatever you want.

**Comment by emilypi** [comment]
@fommil can we get a merge for this since @fosskers proved there were no regressions? We can submit license changes + updates in separate PR's.

**Comment by TomasMikula** [comment]
My gut feeling is that we are not missing out much on improvements being contributed back: for a company, wouldn't it be cheaper to contribute improvements back than to maintain a private fork?

I see how LGPL might be more appealing to contributors. OTOH, it might scare some potential adopters (even though perhaps irrationally). I favor contributors.

If asked, I would agree to relicense my contributions to scalaz 8 under LGPL. But I don't think it makes that much of a difference to be worth the hassle.

**Comment by varming** [comment]
Perhaps this issue is dead already, but I figured I might as well give my opinion.

TL;DR, if you switch to LGPLv3.0, then I will not be able to use Scalaz, thus I am -1 on this change.

Inside large corporations third party software imports are often approved or rejected based on the license of the software and the licenses of the general corpus of third party software previously imported. Where I work lots of GPLv2 code has been imported. Same for LGPLv2.1 and more liberal licenses. Some licenses are prohibited with no exceptions: GPLv3, AGPL, etc. The problem with LGPLv3.0 is that it is not compatible with GPLv2. Both licenses have upgrade clauses that enable uses of the software under the GPLv3 license, so when you link GPLv2 code with LGPLv3.0 code the result is often GPLv3, but GPLv3 is prohibited with no exceptions. As a consequence of this the LGPLv3.0 license is prohibited with no exceptions.

But what if you really want to use LGPLv3.0. Other important libraries have chosen to use LGPLv3.0, e.g., libGMP. Those libraries typically publish code under both LGPLv3.0 and GPLv2 licenses. When that happens I can still use the library under GPLv2.

**Comment by tonymorris** [comment]
@varming Scalaz will not be switching to LGPL. You can tell, because then it wouldn't be Scalaz anymore. It would be something. Take it easy.

**Comment by fosskers** [comment]
Shall we close this issue then?

**Comment by fommil** [comment]
@varming that's an interesting position your company is taking, I've not heard of any scala company accepting GPLv2 but not GPLv3! Note that GPLv2 is also incompatible with Apache 2.0, so you might have bigger problems, given that Scala is now Apache 2.0.

When I contribute to scalaz-core I will use whatever license everybody decides upon, but I'm going more copyleft for my own spinoffs and experiments (that reminds me, I need to find a good Haskell license I'm comfortable with, as I'm not interested in Scala anymore, and I don't think LGPL works well in a language that statically links).

BTW, I personally wouldn't be happy starting a project with "LGPL version 2 or any later version", because GPLv2 is flawed. I only use GPLv2 if I am adapting something that uses that license, e.g. something based on the openjdk sources.

I vote to close. I would be delighted if we used LGPLv3 but only if it is completely unanimous, and that is clearly not the case.

That said, I do think we need to change the copyright owner list to reflect who is actually pushing code in scalaz8. Note that this does not include me :smile: (yet)

**Comment by NeQuissimus** [comment]
+1 to close.

Note that on the microsite I have been using "Scalaz contributors" rather than the list of names. That may be more viable...

**Comment by jdegoes** [comment]
Please don't close. I'd at least like to switch to Apache 2 in Scalaz 8. I think there's probably unanimous agreement from active contributors.
